Add DevServices support for Keycloak

sberyozkin opened the pull request:
Fixes #17217 

CC @stuartwdouglas @pedroigor @stianst

This is an initial prototype which I did by copying from the similar existing main code and PRs. 
At the moment all it does it starts a Keycloak image - from here on it would be all about getting the loaded Keycloak instance prepared as discussed at #17217 as well as creating a Keycloak specifc Dev UI page/card (and as suggested there I think it would need to be moved into a dedicated `dev-services-keycloak` extension since not all `quarkus-oidc` users are KC users).

I've tried to verify this initial code with `quarkus-quickstarts/security-openid-connect-quickstart` (beforee starting it I removed everything from `main/resources/application.properties` and commented out `QuarkusTestResource(KeycloakServer.class)` in `BearerTokenAuthenticationTest`).

So I did:
`mvn quarkus:dev` - and what a magic, I could see KC image was loaded successfully - and when I type `localhost:8081/auth` - Keycload Admin console is ready and I login as `admin:admin`, all is good. There for a start I imported the demo realm - so at this moment KC is ready for the tests to run again.
But when I press `r` - `DevServicesProcessor` does try to launch Keycloak again - @stuartwdouglas the code there which controls whether the docker image has to be restarted looks fine, I could not spot anything wrong there.

So the  question is, is `r` (resume the tests) actually reloads the whole endpoint which I guess would explain why `DevServicesProcessor`  does not recognize that KC is already running.

But it will be fun - we can add a lot of interesting features there around security...

sberyozkin committed:
Add DevServices support for Keycloak

stuartwdouglas commented:
@sberyozkin this is by design, you get one service for tests and one service for dev mode, so your test service does not interfere with you dev mode one. For databases this stops changes you make in dev mode from breaking your tests (and visa versa).

geoand commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DevServicesConfig.java:
```diff
@@ -0,0 +1,60 @@
+     * The Keycloak realm path.
+     */
+    @ConfigItem
+    public Optional<String> realm;
```
I think this property needs more explanation, for example is it a file system path, a classpath path, both?

sberyozkin replied:
@geoand Yes, it was only a quick prototype, for the moment it is treated as a classpath resource but a file system path will be supported as well. For a while it will be me trying to figure how Dev Services work :-)

geoand replied:
Cool 😎

sberyozkin replied:
@geoand I've prototyped the code to support a host system resource as well

sberyozkin commented:
@stuartwdouglas 

> this is by design, you get one service for tests and one service for dev mode, so your test service does not interfere with you dev mode one. For databases this stops changes you make in dev mode from breaking your tests (and visa versa).

Hi Stuart, I've been looking at it like this: I start `quarkus:dev` without KC running, tests fail, then I prepare KC with Dev Services, resume the tests, they pass. It seems quite intuitive to me.
 
So if Keycloak launched with Dev Services is not there to backup the tests then why have it starting with Dev Services ? 

Do you mean that for example in `BearerTokenAuthenticationTest` I do keep `@QuarkusTestResource(KeycloakServer.class)` which launches the test Keycloak instance, while the one we launch with `Dev Services` is for independent improvement of the service endpoint/OIDC configuration/etc ? But then why pressing `r` returns to `DevServicesProcessor` and launches another Keycloak ?

(as a side note I think the `port` configuration property to support a fixed port should be removed then in this PR and other DevServices code which rely on a similar `port` property if `DevServices` is expected to launch more than one image)

I think the flexibility of having one KC per tests, one per Dev Mode can be interesting in some cases - but perhaps it should be optional ?

sberyozkin commented:
I think for now I'll focus on moving it into a dedicated extension as the next step for the actual KC setup improvements is to register the realm, etc using the admin client API

stuartwdouglas commented:
No, I mean that for all the other services we have DevServices for you want two different services. If the tests rely on specific data in the database any changes you have manually made in dev mode could fail the tests. Similarly the tests will often clean the DB which would interfere with anything you were doing in dev mode. For KC this is less of an issue, and could probably be solved via seperate realms.

I will look into this when I am back from PTO, for now just stick with the current approach. It is actually more complex to share them than it is to just have them seperate.

sberyozkin commented:
Hi @stuartwdouglas 
> For KC this is less of an issue, and could probably be solved via seperate realms.
Yes, possible option.
> I will look into this when I am back from PTO, for now just stick with the current approach.

Sounds good and enjoy PTO :-)

For now I'll just move it into a dedicated extension and add an initial realm/client/users registration code.
thanks

sberyozkin commented:
I've created a new `quarkus-devservices-keycloak` which will have its own page.
`quarkus-oidc` still has its `deployment-spi` module for someone else to create the pages for other providers if it can be of interest. 
The other thing is that `quarkus-oidc` has `embedded.html` so that if `quarkus-devservices-keycloak` is available then the OIDC card will look like this:

![oidcKeycloak](https://user-images.githubusercontent.com/467639/119173525-48255e00-ba5f-11eb-9a43-afe6f6cfab28.png)

Note `Provider: Keycloak` is a link to a Keycloak page (or if someone will have for ex Auth0) then it will have `Provider: Auth0` with a link to the `Auth0` page.

At the moment it does not work correctly, since `quarkus-oidc` and `quarkus-devservices-keycloak` are 2 different extensions, so what happens now is:
- `quarkus-oidc`/`embedded.html` has `urlbase` pointing to `quarkus-oidc` based page - while it needs to point to the `quarkus-devservices-keycloak` based page. I guess `quarkus-devservices-keycloak` needs to provide something like `providerurlbase` for `quarkus-oidc`/`embedded.html` to refer to it
- `quarkus-devservices-keycloak` will provide a provider name `Keycloak` but `quarkus-oidc`/`embedded.html` can not refer to it using the `info` namespace since it is provided by the different extension.

So it is about different devservices extensions exchanging the information between each other... I'm not sure yet how to address it, CC-ing to @phillip-kruger as well

thanks.

(But it is just awesome what is already possible with Dev Services)

sberyozkin commented:
It occurred to me both issues can likely be addressed with a dedicated build item provided by `quarkus-oidc/deploment-spi`

phillip-kruger commented:
Hi @sberyozkin - sorry for the late reply. Not sure if I follow the issue properly, so I might be talking nonsense, but if you want a frontend/ui only solution, maybe you can use javascript events ? Not sure if that is a good idea though...

sberyozkin commented:
Hi @phillip-kruger Np at all, thanks for the idea; I may ping you a bit later to clarify how it might work - for the moment I'm experimenting with the idea of using a `DevServicesRuntimeBuildItem` for the OIDC provider specific extension to `inject` some data into a `quarkus-oidc` namespace.

sberyozkin commented:
Hi Phillp, I'll definitely ask you for some JQuery hints as for the moment the KC page is looking a bit sad :-), 
but I've managed to get the properties set by `dev-services-keycloak` available to `quarkus-oidc`/`embedded.html` using `DevConsoleRuntimeTemplateInfoBuildItem` (it is really a built time info but this item is the only one for now which allows to contribute to the other extension's namespace), I'd suggest to enhance `DevConsoleTemplateInfoBuildItem` to support the same, lets see what Stuart thinks about it, CC @stuartwdouglas.

Next, I'm going to experiment with the actions

sberyozkin commented:
Hi @phillip-kruger Quick question, can you see why this [action](https://github.com/quarkusio/quarkus/pull/17364/files#diff-1386362891d85ff4315fc9eecbe0c6dd8f7c9e7478dedeacffd4c0638b02e088R32) is not activated, I keep pressing on it and nothing happens, I prototyped this [action handler code](https://github.com/quarkusio/quarkus/pull/17364/files#diff-a1990f5788485562de0cf2935f6f2612a82caeaf70fdeaecd4937229a5f7f887R46) which is supported by [this build item code](https://github.com/quarkusio/quarkus/pull/17364/files#diff-279ed5b7dab02e019969268179fd5807e10f3d75a3cbcbd5468af84c116a4e4aR36)

phillip-kruger commented:
Hi @sberyozkin - I'll have a look in a bit and get back to you.

phillip-kruger commented:
@sberyozkin the javascript side looks ok (same as this example that is working: https://github.com/quarkusio/quarkus/blob/3ebd9504bc835be6ce28276ef5f840f6c323e790/extensions/vertx-http/deployment/src/main/resources/dev-templates/io.quarkus.quarkus-vertx-http/config.html#L141) Is there any output in the browsers web console ? There you should see if the function is called and if a network request is being made, and what the response was ? Or nothing there ?

sberyozkin commented:
@phillip-kruger thanks for having a look, will continue tomorrow, will check the `config.html` resource as well; I did not see any output, well, I periodically see `Response was already written` but will verify it is not related; cheers

phillip-kruger commented:
Yea the response already written is also a problem. We need to look at that. I have also seen that often. I'll have a better look today, getting the code running on my laptop, as yesterday I just looked at the code

sberyozkin commented:
Here is the current status:
- if `quarkus-devservices-keycloak` dependency is added then the `OIDC` card will have a `Provider: Keycloak` link embedded pointing to a Keycloak page as shown above. Without this dependency the `OIDC` card will only point to the docs as is the case now. 
- If no `keycloak.version` property is set then the `latest` Keycloak image will be used
- will be easy to start an image with the KC preview services enabled if needed later on 
- if `quarkus.keycloak.devservices.realm-path` points to the `existing` class or file system resource then KC will be initialized with this realm file. If the realm file does not exist then by default a `quarkus` realm, a `quarkus-app` client, `alice` and `bob` users, with the `user` role will be created - all of it is configurable.
- One can `Sign in to Keycloak` by clicking a link on the Keycloak page - and create the new realms/clients/users and export the realm to a configured path - so after the restart the realm file will be used.

@stuartwdouglas, 
I've looked again at the problem of the Keycloak image being restarted if the Continuous Testing `r` option is pressed.
I propose to keep the same image instance for the testing and the dev mode work in general for now.  Being to start running tests against the Keycloak instance just being prepared in a dev services KC page would be a good start.

The next KC page design  problem is what users can do there beyond clicking on a `Sign In To Keycloak` link.
I've added a `CreateToken` action (thanks to @phillip-kruger  clarifying the rules about the handlers) just to check an access token can be obtained and indeed it works,

I think the following would be useful enough:
- offer to create a token 1) for a user name which can be selected from a combo box (with `quarkus.keycloak.devservices.users` property feeding it) or manually entered (this can be useful if more users have been added directly in the KC admin console) and 2) for this token be sent to a relative path to be added manually - for example, we will know that the Quarkus endpoint is running for example at `localhost:8080`  but the relative JAX-RS paths are not known so the user would enter a path value like "myservice"

So I guess the user would have a `Test service` button - and once this button is pressed - the chosen user name is used to obtain a JWT token from KC (this already works) and send the token to `http://localhost:8080/{configuredPath}` and show the response (for ex display 200 or 401 or 403 status)

sberyozkin commented:
I suggested above, `KeycloakdevServicesRecorder` now 1) acquires a token from Keycloak using KC Builder API 2) Uses Vert.x Client to send this token to the endpoint with `provider.html` getting the required user and the endpoint relative path (I'll ask Phillip for some help with making it more presentable :-) ).

The problem with this call sequence is that I'm seeing 2 errors in the log:

```
 2021-05-28 18:10:46,162 INFO  [io.qua.dev.key.run.KeycloakDevServicesRecorder] (security-openid-connect-quickstart-dev.jar) (DEV Console action) Test token:  eyJhbGciOiJSUzI1NiIsInR5cCIgO....0KVfCae7b10PglF4vBoQIR2kAzWOo4M0Vy-CTRAEA

2021-05-28 18:10:46,205 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (security-openid-connect-quickstart-dev.jar) (DEV Console action) HTTP Request to /q/dev/io.quarkus.quarkus-devservices-keycloak/testService failed, error id: 2d987c44-b9bb-4292-a48e-66bb4690f23c-1
java.lang.IllegalStateException: Response head already sent
      at io.vertx.core.http.impl.Http1xServerResponse.checkHeadWritten(Http1xServerResponse.java:663)
      at io.vertx.core.http.impl.Http1xServerResponse.setStatusCode(Http1xServerResponse.java:144)
      at io.quarkus.devservices.keycloak.runtime.KeycloakDevServicesRecorder$3.actionSuccess(KeycloakDevServicesRecorder.java:74)
      at io.quarkus.devconsole.runtime.spi.DevConsolePostHandler$2.run(DevConsolePostHandler.java:61)
.....

and

```
 2021-05-28 18:10:46,209 ERROR [io.ver.ext.web.RoutingContext] (security-openid-connect-quickstart-dev.jar) (DEV Console action) Unhandled exception in router
java.lang.IllegalStateException: Response has already been written
      at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:394)
      at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:382)
      at io.vertx.core.http.impl.Http1xServerResponse.end(Http1xServerResponse.java:361)
      at io.quarkus.vertx.http.runtime.QuarkusErrorHandler.handle(QuarkusErrorHandler.java:121)
      at io.quarkus.vertx.http.runtime.QuarkusErrorHandler.handle(QuarkusErrorHandler.java:24)
      at io.vertx.ext.web.impl.RouteState.handleFailure(RouteState.java:1133)
```

So by the time the handler's `actionSuccess` is called the response has already been sent - may be it is related to the other one

sberyozkin commented:
By the way, the idea of setting an endpoint relative address would be better be realized via for ex Swagger UI - where it would only apply a token acquired from Keycloak, but I'd say it is better to do in the next iteration

sberyozkin commented:
`java.lang.IllegalStateException: Response has already been written` has gone now that Georgios has fixed `QuarkusErrorHandler` but

```
HTTP Request to /q/dev/io.quarkus.quarkus-devservices-keycloak/testService failed, error id: fdf74bb0-0a91-4dd6-b5fb-5787cf5b0d7a-1: java.lang.IllegalStateException: Response head already sent
	at io.vertx.core.http.impl.Http1xServerResponse.checkHeadWritten(Http1xServerResponse.java:663)
	at io.vertx.core.http.impl.Http1xServerResponse.setStatusCode(Http1xServerResponse.java:144)
	at io.quarkus.devservices.keycloak.runtime.KeycloakDevServicesRecorder$3.actionSuccess(KeycloakDevServicesRecorder.java:74)
	at io.quarkus.devconsole.runtime.spi.DevConsolePostHandler$2.run(DevConsolePostHandler.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

remains. 

`DevConsolePostHandler.dispatch` calls  `handlePostAsync` (`handlePost` can be used since KeycloakDevServicesRecorder has to block waiting for a response from the service)

sberyozkin commented:
Joining on that thread does not help, strange

sberyozkin commented:
Silly me, `actionSuccess` was called twice

sberyozkin commented on extensions/devservices/keycloak/deployment/src/main/resources/dev-templates/provider.html:
```diff
@@ -0,0 +1,53 @@
+          client: "quarkus-app",
+          user: "alice"
+        },
+        function(data, status){
```
@phillip-kruger Hi Phillip, how to set correctly the value of the result field with the id `result` ? I've tried `$('#result').prop(data);` and just `$('#result').prop("OK");` without the `if` precondition.

phillip-kruger replied:
the element with id `result` is an input, so maybe try `$('#result').val(data);`

phillip-kruger replied:
PS: I have not tried it, so might not work

sberyozkin replied:
Thanks Phillip, I'll try it

sberyozkin replied:
@phillip-kruger Yes it works, thanks

sberyozkin commented:
This is how the Keycloak page looks like at the moment:

![KeycloakDevServcie](https://user-images.githubusercontent.com/467639/120641176-33879380-c46b-11eb-9c0e-60d44252bc30.png)

The user can sign in to Keycloak and further update the realm or start testing the endpoint - given a user name, a token will be acquired and sent to a service endpoint with a relative path set in this page.

What I'd like to try as well, that if it is a `quarkus.oidc.application-type=web-app` then instead offer an option to sign in to the service endpoint.

And I'll start working on the docs - I think it will need to be a dedicated doc since we have one doc for the service application type, one for the web-app type. So both of these docs will refer to the dev services doc.

One thing I'd like to explore is about linking to Swagger UI as an alternative to a primitive (current default) option as shown in the attached image to test the service.

For example, if `quarkus-devservices-keycloak` detects a Swagger UI capability then instead of the above there would be something like:

```
User: (this is needed to get a token from KC)

# but as an alternative to Path/TestService/Result it will be
Test Service With Swagger UI
```

The user presses it and gets a Swagger UI page initialized with a token acquired for a configured user (this token acquisition is already working so will be easy to do) and then the user will test the methods from a Swagger UI.

@phillip-kruger Do you think it can be possible to wire this option in ?

stuartwdouglas commented on extensions/devservices/keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/DevServicesProcessor.java:
```diff
@@ -0,0 +1,393 @@
+
+        boolean needToStart = !ConfigUtils.isPropertyPresent(AUTH_SERVER_URL_CONFIG_KEY);
+        if (!needToStart) {
+            LOG.debug("Not starting devservices for Keycloak as 'quarkus.oidc.auth-server-url' has been provided");
```
This needs to be above the docker warning.

phillip-kruger commented:
@sberyozkin - re swagger-ui, I think that should be possible.

sberyozkin commented:
@stuartwdouglas @pedroigor I think the first iteration is mostly complete, as far as I'm concerned it is nearly ready to go. There is much more we can try to do but I'd say let the users try what is already available. (See the added docs as well)
I'll open an issue immediately after it has been merged for us with Phillip to explore how the acquired test token can be fed into Swagger UI page - it would be quite cool if it works out.
Stuart, one thing this PR still needs to address is that problem with pressing the `r` (continuous testing) which tries to launch a new docker instance, IMHO for now the instance should be reused - it feels intuitive to me, start KC with Dev Services - prepare it, start adding code, test from the Keycloak Dev Services Page, then start adding tests, all running against the same KC.
Then as far as the tests are concerned, users may prefer to use Wiremock or `TestSecurity` etc.
What do you think ? I'd consider adding it more flexible (multiple containers) when it becomes a bit more obvious why it can be needed in case of KC :-)

stuartwdouglas commented on docs/src/main/asciidoc/security-openid-connect-dev-services.adoc:
```diff
@@ -0,0 +1,139 @@
+2021-06-04 16:23:11,157 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
+----
+
+The `quay.io/keycloak/keycloak:latest` Keycloak image is used to start a container. The `quay.io/keycloak/keycloak` image version can be changed by setting a `keycloak.version` property.
```
We can't use the `:latest` version, as it results in un-reproducable builds/tests. You don't want you CI to suddenly break because a new version of Keycloak has been released.

stuartwdouglas commented on docs/src/main/asciidoc/security-openid-connect-dev-services.adoc:
```diff
@@ -0,0 +1,139 @@
+
+You do not need to configure `quarkus-devservices-keycloak` or `quarkus-oidc` to start developing your Quarkus `OIDC` applications with the only exception is that `quarkus.oidc.application-type=web-app` has to be set in `application.properties` to give the `Keycloak` page a hint it needs to show an option to `Sign In To Service`.
+
+By default, the `quarkus`, `quarkus-app` client with a `secret` password, `alice` and `bob` users (with the passwords matching the names), and `user` and `admin` roles are created, with `alice` given both `admin` and `user` roles and `bob` - the `user` role. This should be enough to get started.
```
I would just go 'user' and 'admin' instead of Alice and Bob. A lot of devs won't have studied crypto and won't get the reference.

sberyozkin replied:
@stuartwdouglas We already use `alice` in all the quickstarts, tests (wiremock test resource manager creates it as well), docs. It is just a default user name we used with Pedro, without any particular references to the security scenarios. I'm OK with changing it to `user`/`admin` but we should probably do it separately to make it consistent across it all (quickstarts/docs/tests).

stuartwdouglas commented on extensions/devservices/keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/DevServicesProcessor.java:
```diff
@@ -0,0 +1,398 @@
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ConfigUtils;
+
+public class DevServicesProcessor {
```
Minor nit, but this should be KeycloakDevServicesProcessor, if every extension just call it this is gets hard to open them in the IDE.

stuartwdouglas commented on extensions/devservices/keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/DevServicesProcessor.java:
```diff
@@ -0,0 +1,398 @@
+    private static final String CLIENT_SECRET_CONFIG_KEY = CONFIG_PREFIX + "credentials.secret";
+    private static final String KEYCLOAK_URL_KEY = "keycloak.url";
+
+    private static final String KEYCLOAK_VERSION = System.getProperty("keycloak.version", "latest");
```
This should be 13.0.1, and we already have a config system, we should not be reading random stuff from system properties when it can already be set via config.

stuartwdouglas commented on extensions/devservices/keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/DevServicesProcessor.java:
```diff
@@ -0,0 +1,398 @@
+        // figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container have already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
```
This should not be restarted for every test run. Looks like some other extensions do this but this is likely due to a potential 'residual data' issue which does not affect keycloak. These other extensions that do this likely need to be fixed as this will significantly slow down continuous testing.

stuartwdouglas commented on extensions/devservices/keycloak/runtime/pom.xml:
```diff
@@ -0,0 +1,56 @@
+          <artifactId>quarkus-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
```
We can't be bringing extra dependencies into runtime to make devservices work. All this sort of stuff needs to be in the deployment module.

sberyozkin replied:
@stuartwdouglas Hi Stuart, i followed an approach taken by the cache extension where the template actions are backed up by the recorder code - but in this case the vert.x and keycloak clients need to acqure/post the token - can be done outsode of the recorder ?

stuartwdouglas commented on extensions/devservices/keycloak/runtime/src/main/java/io/quarkus/devservices/keycloak/runtime/DevServicesConfig.java:
```diff
@@ -0,0 +1,125 @@
+import io.quarkus.runtime.annotations.ConfigItem;
+
+@ConfigGroup
+public class DevServicesConfig {
```
This can be moved to deployment, there is no reason for it to be runtime.

stuartwdouglas commented on extensions/devservices/keycloak/runtime/src/main/java/io/quarkus/devservices/keycloak/runtime/DevServicesConfig.java:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * The container image name to use, for container based DevServices providers.
+     */
+    @ConfigItem
```
The default value should be set here.

stuartwdouglas commented on extensions/devservices/keycloak/runtime/src/main/java/io/quarkus/devservices/keycloak/runtime/KeycloakDevServicesRecorder.java:
```diff
@@ -0,0 +1,106 @@
+public class KeycloakDevServicesRecorder {
+    private static final Logger LOG = Logger.getLogger(KeycloakDevServicesRecorder.class);
+
+    public Supplier<String> getProviderName() {
```
All of this stuff can just be done directly in the processor on the -deployment side.

sberyozkin replied:
As far as I recall it was not working

stuartwdouglas commented on extensions/devservices/keycloak/runtime/src/main/java/io/quarkus/devservices/keycloak/runtime/KeycloakDevServicesRecorder.java:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    public Handler<RoutingContext> testService(Supplier<Vertx> vertx) {
+        return new DevConsolePostHandler() {
```
This handler should be in the deployment side, not in the runtime.

sberyozkin replied:
@stuartwdouglas, @cescoffier explained how to get a `Vertx` instance in the deployment, but for now I'm going to focus on replacing the `keycloak-authorization` code with Vert.x code - I already updated the token acquisition code and next I'll have to try and do the realm creation with Vert.x Web Client - hopefully that should be simple since Keycloak `RealmRepresentation` should be convertable to JSON using a Jackson mapper - it it works out then it will be possible to avoid the runtime updates; will look into it after PTO

stuartwdouglas reviewed (changes requested):
I think all this should just go into the quarkus-oidc module, for the following reasons:

- If the runtime dependencies are fixed then this will add no runtime deps on keycloak, so applications written against other OIDC providers will not be affected
- We are allowed to be opinionated, and KC is our preferred solution, so it is ok to make things easy for KC users. It's likely there will not be any other additional devservices implementations for this, but if there is ever a demand for it we can make it pluggable (so KC is default, unless  another one is provided)

sberyozkin commented:
Hi Stuart, @stuartwdouglas, thanks for the review

> I think all this should just go into the quarkus-oidc module, for the following reasons:
>  If the runtime dependencies are fixed then this will add no runtime deps on keycloak, so applications written against other OIDC providers will not be affected

I don't understand it - `keycloak-authorization api` can be added for free, no jar size increase for non KC users ? I could've written a plain client code getting a token, but registering reams/users can be messy without the Keycloak API support.

>  We are allowed to be opinionated, and KC is our preferred solution, so it is ok to make things easy for KC users. It's likely there will not be any other additional devservices implementations for this, but if there is ever a demand for it we can make it pluggable (so KC is default, unless another one is provided)

If someone works with Azure and starts the dev services and sees a link to Keycloak then it will not be great. `quarkus-oidc` is meant to be an OIDC neutral provider, so if it is neutral at runtime but not in the Dev Services view then it will not be great either.
IMHO adding an extra dependency is not a big effort. I appreciate it is not a total Zero effort experience for the Keycloak users, but IMHO we can minimize it to `-Pdev-services-keycloak` - this profile would be declared in `application.pom` and will bring this dependency - would it work for you ? 

And I also doc-ed some suggestions for the users how to support their custom providers - I was thinking that would make Dev Services concept much more interesting to non-Keycloak users - there is nothing really complex about it. Dropping it all now and collapsing it back to `quarkus-oidc` does not sound great to me

sberyozkin commented:
@stuartwdouglas I thought more about it. IMHO doing it pluggable from the get go is correct. We can immediately start suggesting to Okta, Auth0, etc experts to provide Dev Services support - perhaps they'd be able to contribute the dev services extensions for their providers to Quarkiverse. They may blog about it. It will be a good story.

stuartwdouglas commented:
devservices all lives in the -deployment side, not the runtime side, which means it does not end up in the final application. If you add  keycloak-authorization to the -deployment module it will not affect the size of the final application.

If you really don't want it in the quarkus-oidc module then we should create a quarkus-keycloak module that depends on quarkus-oidc, and tell keycloak users to depend on this. It can contain all the Dev UI/DevServices keycloak bits, even though it just uses quarkus-oidc under the hood.

sberyozkin commented:
@stuartwdouglas Hi Stuart,

> If you really don't want it in the quarkus-oidc module then we should create a quarkus-keycloak module that depends on quarkus-oidc, and tell keycloak users to depend on this. It can contain all the Dev UI/DevServices keycloak bits, even though it just uses quarkus-oidc under the hood.

I don't want to be difficult :-), but IMHO it will not work well:
- Keycloak users can use both `quarkus-oidc` and `quarkus-keycloak` to do exactly the same thing at runtime - but without any Keycloak add-ons - but if they start switching to `quarkus-keycloak` they will start asking - why can't we have this Keycloak specific feature - etc - but recall, the reason the original `quarkus-keycloak` was replaced with `quarkus-oidc` was because we agreed with the Keycloak team it would be easier to maintain a neutral extension based on the OIDC standards only. So adding `quarkus-keycloak` will start pushing us back to adding various Keycloak add ons
- The only reason `quarkus-keycloak` would be introduced is for Keycloak users not to type `-Pdev-services-keycloak` - so effectively it won't be different to the PR approach - instead of running `quarkus-oidc` with `-Pdev-services-keycloak` they'd have to make a choice - I can run `quarkus-oidc` or if I also want to have `Dev services for Keycloak ` then I have to run `quarkus-keycloak` - IMHO this is actually a worse option since as I said above we won't be doing any KC specific runtime enhancements.
- It won't really align with the idea of Okta/etc developers adding their dev services support - they'd have to have similarly named `quarkus-okta` (etc) etc.

IMHO it is not a big problem if we doc that to have a Dev services support for KC you need to enable a single -P profile.

stuartwdouglas commented:
We are not doing the profile idea, if every extension started doing things like that you would have a command line a mile long, its not very good from a usability point of view. It's also not really compatible with running directly from the IDE.

sberyozkin commented:
Hi @stuartwdouglas 

> We are not doing the profile idea, if every extension started doing things like that you would have a command line a mile long

The dev services profile idea only makes sense for `quarkus-oidc` which is OIDC provider neutral and hence this concern of a long cmd line will not materialize.

> It's also not really compatible with running directly from the IDE.
I guess in the IDE the actual module dependency would be present ?

OK. So you suggested to introduce `quarkus-keycloak` which will just extend `quarkus-oidc` with the dev services code/templates, as an alternative to inlining the dev services code/templates in `quarkus-oidc`. I'm trying to advocate a standalone `quarkus-dev-services-keycloak` - it does not have a `quarkus-oidc` dependency right now. 

It has occurred to me that if I add the `quarkus-oidc` dependency to `quarkus-dev-services-keycloak` then it will just become a matter of finalizing this extension name, `quarkus-keycloak` or keep it as  `quarkus-dev-services-keycloak` :-)

I can indeed name it as `quarkus-keycloak` and note in the OIDC dev services doc that Keycloak users should use `quarkus-keycloak` to avail of the dev services for keycloak. Would `quarkus-dev-services-keycloak` be a better name though or you'd like to avoid it ?

sberyozkin commented:
@stuartwdouglas It is getting a bit closer to some completion - I've resolved a few of your comments, 2 problems exist at the moment - the runtime code is needed for now and the name of the extension needs to be finalized...

sberyozkin commented:
@stuartwdouglas I've moved most of the code to the deployment, with only [this one](https://github.com/quarkusio/quarkus/pull/17364#discussion_r653142101) remaining in the runtime (please see my comment above why I have to keep it there for now), it is hopefully temporary but it also a very small amount of code requiring no extra dependencies.

I've also renamed the extension, after thinking for a while to `quarkus-oidc-devservices-keycloak` with the description `OpenId Connect Adapter with DevServices Support for Keycloak` - as you can see its runtime only depends on `quarkus-oidc`. I was hesitating to name it as `quarkus-keycloak` - as the only Keycloak specific aspect which will be there is the Dev Services part.
It feels this naming pattern is reasonable, we can then open issues to have `quarkus-oidc-devservices-okta` (`-auth0`) etc and invite the experts to contribute.

As I said once this PR is merged we can start discussing with Phillip how to wire Swagger UI into it for the users not to enter the service paths manually. 

I don't know what else can be done with this PR at this stage.

Note, I've also opened #18427 for `quarkus-oidc` to offer a generic OIDC debugger support.

sberyozkin commented:
@stuartwdouglas We've been looking with @phillip-kruger at the latest PR updates - where a user can get a token pasted into the Dev Console after signing in to Keycloak and then this token is posted to Quarkus. Phillip also liked your idea of having it all done inside `quarkus-oidc` alone - and I clarified we already have a demand to support a generic OIDC Dev Services experience - Phillip thought about making it configurable - I think it is a good idea, let me summarize again the current status and propose how to continue with `quarkus-oidc` only:

- `quarkus-oidc` is not tied up to Keycloak therefore for the users working with for ex `Auth0` and doing `quarkus:dev` and seeing a Keycloak container starting up would be confusing, of course, they can disable keycloak-dev-services next, but I don't think they should have to do it or even be aware that `quarkus-oidc` can start a Keycloak container.
- OOB, `quarkus-oidc`, should provide a limited Dev Services experience which will work for all OIDC providers: #18427. It will assume the provider is already running, Keycloak or a free dev Auth0 tenant as discussed with Jim Tyrrell, etc.
- The documentation will recommend Keycloak users to set a `quarkus.oidc.keycloak-devservices.enabled=true` in `application.properties` or as a `-D` command line property - in which case a Keycloak specific dev experience will be initiated - a KC container started etc.

So we can have `quarkus-oidc` shipping both the OIDC provider neutral and Keycloak specific Dev Services support, OIDC provider neutral experience is enabled by default, `quarkus.oidc.keycloak-devservices.enabled=true` will activate a KC specific support. IMHO this is the best compromise which we can get. 

Stuart, can you confirm please you agree with it ?

stuartwdouglas commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesUtils.java:
```diff
@@ -0,0 +1,74 @@
+        props.add("grant_type", "password");
+
+        return request.sendBuffer(OidcCommonUtils.encodeForm(props)).onItem()
+                .transform(resp -> getAccessTokenFromJson(resp)).await().indefinitely();
```
This should likely have some form of timeout. No sure if you can set a request timeout in the vert.x client itself?

sberyozkin replied:
@stuartwdouglas I could not find it in `WebClientOptions` but this timeout is now configurable and I've updated the code to refer to it

stuartwdouglas commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesUtils.java:
```diff
@@ -0,0 +1,74 @@
+
+        props.add("grant_type", "client_credentials");
+
+        return request.sendBuffer(OidcCommonUtils.encodeForm(props)).onItem()
```
Same here

stuartwdouglas commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java:
```diff
@@ -0,0 +1,395 @@
+        // figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container have already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
```
We should not be restarting for test mode, unless there has been a config change. This makes continuous testing super slow if you need to restart the container every time.

sberyozkin replied:
Hi Stuart, @stuartwdouglas, this is how it is done now:
```
// This step runs for both Test and Developement modes
@BuildStep(onlyIfNot = IsNormal.class)
public DevServicesConfigBuildItem startKeycloakContainer(...) {

        DevServicesConfig currentDevServicesConfiguration = config.devservices;
        // figure out if we need to shut down and restart any existing Keycloak container
        // if not and the Keycloak container have already started we just return
        if (closeables != null) {
            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT
                    && !currentDevServicesConfiguration.equals(capturedDevServicesConfiguration);

            if (!restartRequired) {
                return null;
            }
            // close all the closeables
            // restart
            StartResult startResult = startContainer(currentDevServicesConfiguration);
           ...
}
```

I've copied `@BuildStep(onlyIfNot = IsNormal.class)` from one of the other dev services processors.

Should it be `@BuildStep(onlyIf = IsDevelopment.class)`, i.e, when running the tests we should expect a test specific container starting with the test resource manager ?

sberyozkin replied:
I think I get it - so the container won't start if `quarkus.oidc.auth-server-url' is set - which is a prerequisite for running the tests so the above should be fine (I hope :-) )

stuartwdouglas commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java:
```diff
@@ -0,0 +1,395 @@
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
+            if (!restartRequired) {
+                restartRequired = !currentDevServicesConfiguration.equals(capturedDevServicesConfiguration);
```
We should also check the realm file for changes, so if the user modifies the configuration this should be taken into account.

sberyozkin replied:
Hi Stuart, @stuartwdouglas - I've been thinking about it and nearly implemented a size check, but then realized that it is not really needed - if the realm file exists then Keycloak will be initialized from it - but Dev UI always works against the live current Keycloak instance - i.e, if the realm has been updated via KC Admin Console and re-exported into the same file which KC was initialized from - it won't change anything since whatever updates that have been applied to this realm are live so DevUI will pick them up indirectly when requesting the tokens or letting the users to sign in to KC.

stuartwdouglas replied:
Remember there are two KC instances, the test one and the dev one. Even if you assume it is only updated by export and not by manual modification the test one would still need to be updated to be in sync with the dev one. 

Also the dev and test ones may have separate files (as you could set a different file name in a %test profile), so this still should check this.

sberyozkin replied:
Hi @stuartwdouglas `manual` modification of Keycloak realm is most likely (if not definitely) is not an option - it can be up to 80K and it is quite a complex JSON - I doubt anyone is attempting to change it manually so I would not be concerned about missing the manual updates which might affect a given KC instance - I think we can safely assume/require that the realm file modifications should only be done from the KC admin console's `export realm` option. Myself and others have tried to trim `realm.json` a few times simply to make it smaller - but this is not done to affect the behavior of KC as such :-), it is just a trial and error attempt to produce a small size file which is easier to read/etc...

Re the test case - this one is easy, `quarkus-oidc` does not support KC specific properties or are you saying that we can have 2 sets of `keycloak.devservices` properties - one for starting a devservices container and another one (in a test profile) for running the tests ? Should we allow it ? (I'd be a bit confused if it were the case).

I think as far as the file check is concerned, the only check which can work is a file size check - that is easy enough to do. I'd be concerned that the devservice users will see the container restarting immediately after exporting the realm from the current KC instance console - which would make the current browser view stale: ex, the user is working with `localhost:12345/auth/realms/admin`, exports the updated realm, continues clicking around in that console, while Quarkus has already stopped the container which backs it up and now it runs at `localhost:67890/auth/realms/admin`

sberyozkin replied:
@stuartwdouglas update, `RealmRepresentation` from `keycloak-core` has no `equals` so if if we were to check for the possible updates it would have to be a file size check. I'd prefer though to wait with it until we see an actual issue with not checking it :-), see above for some concerns as well

stuartwdouglas replied:
Of course we can have different properties for running the tests. If you have a `%test.quarkus.oidc.devservices.realm-path` you can use a different file for tests. We don't need a .equals() check for modification all you need to check is if the last modified time has changed. 

Quarkus won't actually restart until you hit the app with the browser. If you were really worried about it you could make the restarted container use the same port.

sberyozkin replied:
@stuartwdouglas Sure, the realm can either be a classpath resource (possibly bundled in a jar) or a file system resource, for the  latter case I can add this check. Good point about the same port in case of restarting the container - it should probably be done anyway - but I'll look at it if there will indeed be some issues related to syncing the views with the running containers

Re running the tests with the alternative devservices setup - not sure it really works in practice though - i can certainly see how I can start building tests while I'm doing the dev against the current KC container - but eventually I'd have tests depending on the test resource factory starting a container (as is the case for all the OIDC tests).. That said, I see, the dev mode tests can indeed depend now on the devservices config... interesting, may be I will try to update Quarkus OIDC dev mode test to try once the code is merged...

stuartwdouglas commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java:
```diff
@@ -0,0 +1,395 @@
+    private ClientRepresentation createClient(String clientId, String oidcClientSecret) {
+        ClientRepresentation client = new ClientRepresentation();
+
+        client.setClientId(clientId);
```
Should this configure CORS to automatically allow requests from localhost:8080 (or whatever the HTTP/HTTPS listeners are bound to)?

I tried to enable OIDC in Swagger UI and I could not figure this bit out, as keycloak would not send back an Access-Control-Allowed-Origins header.

sberyozkin replied:
@stuartwdouglas Hi Stuart - do you mean you could not authenticate to Keycloak via Keycloak's authorization endpoint ? That endpoint does not support CORS - I'd propose to look at wiring in Swagger UI in a follow up task, such as #18597

stuartwdouglas replied:
Why doesn't it support cors? I thought keycloak.js would send cross origin requests to this endpoint to perform authentication?

sberyozkin commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesUtils.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public static WebClient createWebClient() {
+        Vertx vertx = Vertx.vertx();
```
@stuartwdouglas I thought that closing the client would close this instance. I can just close it at the same time the client is closed, that would be simpler ?

sberyozkin replied:
This is done as per your recommendation - let me know if it needs to be tweaked please

sberyozkin commented on bom/application/pom.xml:
```diff
@@ -623,6 +623,16 @@
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-oidc-devservices-keycloak</artifactId>
```
@stuartwdouglas 
`quarkus-oidc-keycloak` sounds good to me, what about the suggestion to have it all in `quarkus-oidc` - please see my comment above.  
So 2 options - we have either 1) `quarkus-oidc` and Keycloak Dev Services/UI are enabled via the property or 2) we have `quarkus-oidc-keycloak`. 

If we have  `quarkus-oidc-keycloak` then it would be closer to a zero effort idea of enabling Dev Services, but I'm not sure now again, I think `quarkus-oidc` only with Keycloak users having to add one property is not bad but now sure if you agree - let me know please - I think both options have their pros and minor cons, I'm fine with either option at this stage, just let me know what you prefer please

sberyozkin replied:
Hey Stuart, @stuartwdouglas, I've come the full circle in my understanding how it should be packaged and after all these tries and knowing I can put it all into `deployment` without affecting the final jar size, your original idea of having it all in `quarkus-oidc` can work well for all the variations. Let me summarize, hopefully this PR will be ready to go soon, so:

1. if `quarkus:dev` is started with the empty `application.properties` then it is reasonable to default to KC specific Dev Services/UI experience as the user has not indicated which provider is required - so OOB the `OIDC`card will link to the `Keycloak` page.
2. The user will disable it by doing `quarkus.oidc.devservices.enabled=false` - I think there should be no `keycloak` in this property name, keycloak specific configuratiions would be in the `quarkus.oidc.devservices.keycloak.` namespace. 
 2.1 If non-Keycloak users will push back against having to disable it to avoid the Keycloak containers starting in their case OOB with the empty properties then I think we will just switch the default value - or we can revisit the `quarkus-oidc-keycloak` idea, lets see how it goes.
3. if `quarkus.oidc.auth-server-url` is already set then the KC container won't start either
4. if no KC container is started then only an OidcDebugger like Dev UI will be provided where the tokens would be acquired from Keycloak/Okta/Auth0/Google/etc using the standard grants/flows
5. if Quariverse will have `quarkus-oidc-okta`/`quarkus-oidc-auth0`/etc dev services extensions then they will override the default OidcDebugger like/ or Keycloak specific Dev Services/UI provided by `quarkus-oidc`
  
I think it all now kind of clicks and should work fine for all the variations, I'll start working on collapsing it all inside `quarkus-oidc` tomorrow morning - unless you'll object by then :-)

thanks

sberyozkin replied:
@stuartwdouglas It is all in `quarkus-oidc` now, I'll be tuning it next

sberyozkin commented on extensions/oidc-devservices-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakBuildTimeConfig.java:
```diff
@@ -0,0 +1,16 @@
+ * Build time configuration for Keycloak DevService.
+ */
+@ConfigRoot
+public class KeycloakBuildTimeConfig {
```
@stuartwdouglas `BuildTime` is not propagating into the property name, `quarkus.keycloak.devservices.` works, I've tested it

stuartwdouglas commented on extensions/oidc-keycloak/deployment/src/main/java/io/quarkus/devservices/keycloak/deployment/KeycloakDevServicesProcessor.java:
```diff
@@ -0,0 +1,410 @@
+        // Figure out if we need to shut down and restart any existing Keycloak container
+        // if not and the Keycloak container has already started we just return
+        if (closeables != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.DEVELOPMENT
```
This should not factor in the launch mode, it should just be based on if the config has changed (but the 'config changed' check should be expanded to include if the json import file changes). We only want to restart the container if it is absolutely necessary, because it is slow.

Note that this behaviour does vary a bit between the DevServices. Mongo and Redis in particular are restarted to make sure there is no left over data. The SQL databases are not as hibernate knows how to clean out all the data between test runs. At some point this restart behavior for the other ones may be improved, because in general you want to restart as little as possible.

sberyozkin replied:
Hi @stuartwdouglas OK, I'll remove the `DEVELOPMENT` check, makes sense; but as I commented [here](https://github.com/quarkusio/quarkus/pull/17364#discussion_r668014434), I believe we don't need to check the realm file changes - since those changes can only happen if the user (re-)exports whatever has been updated in the live Keycloak instance's console - but these changes are indirectly picked by the Dev UI which requests the tokens from KC.

stuartwdouglas commented on extensions/oidc/deployment/src/main/java/io/quarkus/oidc/deployment/OidcBuildTimeConfig.java:
```diff
@@ -13,4 +14,10 @@
+     * Dev services configuration.
+     */
+    @ConfigItem
+    public DevServicesConfig devservices;
```
This config is keycloak specific, shouldn't it be under quarkus.keycloak rather than oidc?

sberyozkin replied:
Hi Stuart, @stuartwdouglas 
Yes, I was thinking about it, Keycloak is a `default` devservices option, hence I thought about avoiding `keycloak`. The docs will say that `quarkus.oidc` will launch a KC container by default in the dev mode but it can be disabled with `quarkus.oidc.devservices.enabled=false`

I'm also looking at how to handle the case where the OidcDebugger only Dev UI is presented (or if a Quarkiverse `oidc-okta` etc is available), i.e, if I'm an Okta user then I feel I'd rather do:
```
# disable the default devservices option:
quarkus.oidc.devservices.enabled=false
# Configure Okta container, etc
quarkus.okta.devservices....
``` 

as opposed to
```
# disable the default devservices option:
quarkus.keycloak.devservices.enabled=false
# Configure Okta container, etc
quarkus.okta.devservices....
```

in this last case a `keycloak` property looks a bit intrusive to me as an OKta user :-) - I'd like to avoid introducing a requirement for non-KC users to refer to Keycloak in their configuartion.

I think that if we, at a later point, either still in this PR of after it has been merged, decide to return to the idea of shipping a dedicated `quarkus-oidc-keycloak` then indeed it would be `quarkus.keycloak.devservices...`.

The alternative is indeed to have `quarkus.keycloak.devservices` namespace within `quarkus-oidc` but disable `quarkus.keycloak.devservices.enabled=false` by default - in this case the only users who will be dealing with `quarkus.keycloak.devservices` will be Keycloak users.

I'm getting a bit dizzy with all these options :-), perhaps I'm over-thinking it. But I believe it is important to keep a 'neutral` aspect of `quarkus-oidc` when it comes to configuring it as well.

So I'd either 1) keep it as is now (since KC is default for devservices, `quarkus.oidc.devservices` namespace if used to configure it) or we do 2) `quarkus.keycloak.devservices` within `quarkus-oidc` but disable by default

Perhaps 2) is not bad - recall, OOB, there will eventually be an `OidcDebugger` like Dev UI experience anyway where one would expect a container started separately or an Auth0/etc account is available, etc. To get a KC devservices experience KC users would have to do `quarkus.keycloak.devservices.enabled=true` - extra effort - but it is comparable to effort which would be required if they had to choose `quarkus-oidc-keycloak` (instead of `quarkus-oidc`) to get this experience.

IMHO, perhaps staying with 1) is OK for now but we should review 2) if you prefer to keep it all under the `quarkus.keycloak.devservices` namespace

let me know what you prefer please, thanks

sberyozkin replied:
@stuartwdouglas I have tried `oidc/deployment` tests with this PR and I've found a few issues related to `quarkus.oidc.devservices.enabled=true` (which is a default):
1. Dev Mode test which starts without `quarkus.oidc.auth-server-url` and verifies the auth request fails actually passes (or rather fails to fail :-) ) since the dev services container is started and this property is provisioned - so I had to disable the devservices to verify it passes
2. if we have a case where OIDC multi-tenancy is tested (without the default config, without `quarkus.oidc.auth-server-url`, which is usually the case when `TenantConfigResolver` is used to create the config dynamically at the request time) then the devservices container will also be starting during the tests
3. Normal non-dev mode security disabled test (which is also run by QE) also spawns a container - test passes - but slowly

For 1. and 3. I could fix the problem by disabling the devservices in the config - for 2. - it is not really fixable - or rather, I don't know if starting the devservices container can be prevented.

All of these issues would stay even if we revert to the idea of having `quarkus-oidc-keycloak` and disabling the devservices there would be worse since it would be an extension designed to offer KC devservices exp out of the box.

Given this and the earlier comment, I'm coming to the conclusion that the best option is to have `quarkus.keycloak.devservices.enabled=false` by default (with the configuration being in the `quarkus.keycloak.devservices` namespace as you suggested).

So when a user will start `quarkus-oidc` with the empty `application.properties` - the empty OIDC card will be shown. Next user goes to application.properties and adds `quarkus.keycloak.devservices.enabled=false` - and the container has started - or the user adds `auth-server-url` - in this case OIDC card will show a link to an OIDC page where one would do an OidcDebugger like view. `quarkus-okta`/`quarkus-auth0` benefit as well - there would be no need to disable the default devservices. I'll prototype it shortly, IMHO this works for all the cases

stuartwdouglas replied:
I think it should be quarkus.keycloak.devservices, if we have a quarkus-okta then we can use build items to stop keycloak being enabled by default (i.e. quarkus-okta would provide a build item that would disable the keycloak devservices).

Basically this means that keycloak is the default and works out of the box, but if you add another extension it turns off keycloak. 

In terms of your test problems:

1) is completly fine, other tests have had this problem, just turn it off for the specific test
2) I think if multi tencancy is enabled the devservices should probably be disabled. Not 100% sure about this though, maybe just turn it off for the test
3) If security is disabled why isn't devservices disabled?

The whole point of dev services is that it is zero config, it should not have to be specifically enabled.

sberyozkin replied:
Hi Stuart @stuartwdouglas, it all makes sense, I've already changed the namespace yesterday, and good idea about disabling KC devservices indirectly via the dedicated build item (this item can also provide some extra info). OK, tests can be taken care of as well.

> The whole point of dev services is that it is zero config, it should not have to be specifically enabled.

Right this is where the design `tension` is :-), yes, I totally agree we'd like the KC users to have the lowest effort applied toward enabling the dev services; but what somehow concerns me is that if you are not a KC user you won't have a zero effort at starting `quarkus:dev` - `quarkus-okta`/`quarkus-auth0` are just the ideas yet to be realized by those provider's experts which may or may not happen - so these users would have to type `quarkus.keycloak.devservices.enabled=false` before starting `quarkus:dev` - they don't have to do it now - while after this PR they will have to and a `keycloak` config root would become a permanent feature of their configuration - `quarkus-oidc` will lose some of its `neutrality` toward offering a provider neutral experience. This is the reason behind my hesitation to enable it OOB.

Let me do the final round of various minor updates, docs, and I'll postpone making `quarkus.keycloak.devservices.enabled=true` a default value till the last moment and then I'll ping you to finalize :-) It is nearly there :-)

stuartwdouglas replied:
If you are not using KC then you will have to configure a URL, same as what you do now, and this will disable DevServices.

If we have support for okta/auth0 then you would just include these extensions and then DevServices would stop starting KC and start the other one instead.

sberyozkin commented:
@stuartwdouglas This is now ready for review - I've enabled the services by default, IMHO we should be open to reviewing it, but for now, apart from some UI cleanup/improvements (which I hope Phillip will help with later on) it is ready.

phillip-kruger commented:
@sberyozkin As discussed, I have updated the UI:

- Now only one page (no more spa.html)
- Ability to copy to clipboard (for tokens)

Here an example of the Implicit page:

![implicit](https://user-images.githubusercontent.com/6836179/126121850-ba1fa521-a68c-49e4-9a7a-70d4cb6d88b6.gif)

sberyozkin commented:
@phillip-kruger Hi Phillip, this is brilliant, I like it, nice transition from a primitive UI I did to a quality UI one can actually enjoy working with :-), @stuartwdouglas - FYI. Later on I'll try to reproduce some of it for a a generic OIDC dev UI experience (so that it works for any provider)

I'll look at updating the docs shortly and it will be ready to go IMHO.
Cheers

sberyozkin commented:
@stuartwdouglas thanks; I'll do a few improvements in a follow up PR. I've proposed to backport - the only change which affects the original source is moving the OIDC build time config from `runtime` to `deployment` where it should have been all the time

geoand commented:
Very nice to see this in! I assume the quickstarts will be updated accordingly?

sberyozkin commented:
Hi @geoand I hope so :-), I'm not sure yet what form it should take when combined with Keycloak Dev Services. By the way I've removed a backport label for now as I've hit a problem

sberyozkin commented:
Guillaume, If backported then please backport only if #18856 merged
